src/Dialect/ONNX/ONNXOps/ShapeHelper.cpp

```diff
@@ -378,49 +380,256 @@ bool ONNXBroadcastOpShapeHelper::hasNoBroadcast(DimAnalysis *dimAnalysis) {
-    SmallVectorImpl<IndexExpr> &operandAccessExprs, bool hasNoBroadcast) {
+    int64_t operandIndex, const SmallVectorImpl<IndexExpr> &loopAccessExprs,
+    SmallVectorImpl<IndexExpr> &operandAccessExprs, bool flattenedInnerDims,
+    bool ruledOutBroadcast) {
+  // Get info.
+  int64_t loopDepth = loopAccessExprs.size();
+  int64_t inputSize = inputsDims.size();
+  int64_t operandRank = operand.getType().cast<ShapedType>().getRank();
+  // Flattened? no more than one loop per dim in output (aka output rank).
+  // Not flattened? one loop per dim in output (aka output rank).
+  if (flattenedInnerDims)
+    assert(loopDepth <= (int64_t)outputRank &&
+           "without flattening, expect one loop iter variable per output rank");
+  else
+    assert(loopDepth == (int64_t)outputRank &&
+           "without flattening, expect one loop iter variable per output rank");
+
   // Emtpy the access expr, just in case.
   operandAccessExprs.clear();
-  // There is this case where we have no broadcast per se, but we have mixtures
-  // of 1xTYPE vs TYPE scalars. Handle this case properly here.
-  uint64_t operandRank = operand.getType().cast<ShapedType>().getRank();
+  // There is this case where we have no broadcast per se, but we have
+  // mixtures of 1xTYPE vs TYPE scalars. Handle this case properly here.
   if (operandRank == 0)
     return success();
 
-  // The hasNoBroadcast pattern can be established by shape inference using
+  // The ruledOutBroadcast pattern can be established by shape inference using
   // DimAnalysis. If that is available, and broadcasting was ruled out, then
   // more efficient code can be generated.
-  if (hasNoBroadcast || (hasUniBroadcasting && operandIndex == 0)) {
-    for (IndexExpr ie : outputAccessExprs)
-      operandAccessExprs.emplace_back(ie);
-    return success();
-  }
+  bool noBroadcasting =
+      ruledOutBroadcast || (hasUniBroadcasting && operandIndex == 0);
 
-  for (uint64_t i = 0; i < operandRank; ++i) {
+  for (int64_t r = 0; r < operandRank; ++r) {
     // Shape helper may pretend 1s, thus adjust dimension index accordingly.
-    uint64_t dimIndex = outputRank - operandRank + i;
-    SymbolIndexExpr dim(inputsDims[operandIndex][dimIndex]);
-
-    // Compute access index based on broadcasting rules.
-    // If all other operand dims are 1, just use the output access index.
-    // Otherwise, emit a select op.
-    bool allOtherInputDimsAreOne = true;
-    for (uint64_t i = 0; i < inputsDims.size(); ++i) {
-      if (i == operandIndex)
-        continue;
-      IndexExpr dim = inputsDims[i][dimIndex];
-      if (!dim.isLiteralAndIdenticalTo(1)) {
-        allOtherInputDimsAreOne = false;
-        break;
+    // Take loopDepth instead of outputRank as loopDepth reflect the
+    // (possible) flattening of the loops.
+    int64_t dimIndex = loopDepth - operandRank + r;
+    SymbolIndexExpr operandDim(inputsDims[operandIndex][dimIndex]);
+
+    bool guaranteedBroadcasting = false;
```

Comment:
Now we have another one about broadcasting, which makes me more confused ... Is this additional `guaranteedBroadcasting` necessary for SIMD?

Reply:
I added an example in the code. Repeating here 
```
      // For example we have 5x?x3xf32 and 5x1x3xf32. We don't know at compile
      // time if the `?` will be 1 or (assuming without loss of generality) 10.
      // Normal code is to generate for `?` for that access dimension:
      //   select(dim==1 ? 0 : loop-var)
      // so that if this access needs to be broadcasted, we will only access the
      // `0` value of this broadcasted value; and if it is not broadcasted, we
      // will access each `loop-var` value for this access.
      //
      // But since all of the other dimensions are '1', they cannot generate
      // a broadcasting situation. Thus, for this access function, we can
      // generate the access `loop-var` regardless of whether the `?` will
      // evaluate to 1 or 10 at runtime. It will be correct no matter which
      // situation we will be in.
```

Reply:
I can see how the name is confusing... let me rename it